### ISSUE TITLE
Running hasura console via CLI in docker container

### ISSUE THREAD

**Issue by codepunkt** [hasura/graphql-engine#2824]
Hello!

I'm trying to run `hasura console` in docker container no#1 and then access the console on port `9695` on my host machine.

The JavaScript of the hasura console application needs access to both the graphql engine (running in docker container no#2) and the migration service spun up by running `hasura console` in docker container no#1.

The URLs that the JavaScript tries to talk to are based on two things:
- The `endpoint` configuration of `config.yaml` in docker container no#1
- The `--address` flag used to start `hasura console` with in docker container no#1, which defaults to localhost when not set.

Because the hasura console application is only available inside docker container no#1 when started with the default `--address` of "localhost", but is not available on the host machine of docker container no#1, i need to start it with `--address 0.0.0.0` to have it bind to all available interfaces so it's not only reachable inside of container no#1, but also on the host machine.

Because the graphql engine is running in docker container no#2, which happens to be set up in a `docker-compose.yml` with docker container no#1 and a third run running postgres, i need to configure the `endpoint` in `config.yaml` to the docker compose service name of the second container, which is `endpoint: http://graphql-engine:8080`.

This is the only configuration where the console actually starts, because it can access docker container no#2 from docker container no#1 and where the started console is then actually reachable by my docker host, because it is bound to all available interfaces.

**The problem**:

The console web application now tries to talk to both `http://graphql-engine:8080` and `http://0.0.0.0` when accessed from the browser of my host machine - both of which are not available.

You can work around the first of those requests by having both container no#1 and container no#2 on the same network, so the graphql engine can be reached with `endpoint: http://localhost:8080` - which also works on the docker host when port `8080` is exposed to the host.

I did not find any workarounds for the `--address` and the resulting requests to `0.0.0.0` though. Is there anything that i'm missing?

Would it be possible to add another configuration option to differentiate between the interfaces the graphql engine is actually served on and the URL that the clientside javascript uses to access it?

**Comment by shahidhk** [comment]
This is a known issue and we have some potential solutions. 

The console served by CLI is designed to be run from a developers machine and not from any sort of webserver.

Will you be able to elaborate on what the use case here is, where you're unable to run the hasura cli locally?

**Comment by codepunkt** [comment]
We have packages that don't install/build under windows, which is why we develop in linux docker containers.

**Comment by shahidhk** [comment]
Are you saying the hasura CLI doesn't work on Windows?

**Comment by shahidhk** [comment]
Or it is just a matter of convenience where you'd prefer everything to be inside docker containers?

**Comment by codepunkt** [comment]
It's a matter of convenience. We wouldn't use Hasura/Postgres if they weren't available in official docker images. Development has to be done in a linux based container anyways, because parts of it don't work on windows.

We could possibly run the hasura CLI on windows, but that'd be a pain to do (no automatic installation, having it connect to graphql-engine running in a container, being forced to link files into the dev container etc).

Aside from that, we get "An attempt was made to access a socket in a way forbidden by its access permissions" by the windows executable and i can't tell why. Might be a windows/network/firewall configuration thing - which i might be able to fix, but which might also be company wide settings?

**Comment by shahidhk** [comment]
Got it. We'll ideate on how we can make the experience smoother. 😀

For migrations you'd have to mount the directory inside the container anyway. 

Do let us know if there's an ideal workflow that you would like see. Helps in our ideation/brainstorming 😄

**Comment by codepunkt** [comment]
It'd probably be okay if there were additional options for the URLs the console's javascript is connecting to, something like `endpoint` and `clientEndpoint`, aswell as `--address` and `clientAddress`.

I know the naming is off, but you get the idea 😉

**Comment by shahidhk** [comment]
Thanks @codepunkt. We have an implementation at https://github.com/hasura/graphql-engine/pull/1780 and we should get to merging it soon.

**Comment by xorander00** [comment]
I'm having a similar issue as well.  In my case, it's the ServiceWorker (/sw.js) that the console is trying to fetch, but the hasura is upstream on a different path, so it 404s.  
  
I haven't delved into it yet, but from some quick searching on this repo, it looks like the console is using CRA?  If so, then you should be able to inject the appropriate PUBLIC_URL environment variable (although I can't remember off the top of my head if this only happens during build-time or if applies at run-time as well).  
  
Another option would be to let nginx configure a header using proxy_set_header, and then read that in hasura and pass it onto CRA.  This approach is more flexible since it keeps that config where it's relevant, in nginx.conf.  
  
Anyway, just my $0.02.  Not a show stopper in my case because it's just the service worker that's not loading, and from the looks of it, it's not mandatory.

**Comment by Sevensidedmarble** [comment]
I'd like to chime in and say I'd love to be able to serve the console from the same docker-compose file as I do the hasura instance itself.

**Comment by BlueSeph28** [comment]
There is a workaround to do this without this pull request or creating a new container by my own?
I've run console manually, but doesn't work.

````
INFO console running at: http://localhost:8081/
````
The port 8081 is binded for my container, but I receive in my browser.

>This page isn’t working localhost didn’t send any data. ERR_EMPTY_RESPONSE

**Comment by shahidhk** [comment]
We're ideating on a simpler solution for this where all CLI APIs are protected by admin secret and all that needs to be done is to expose them correctly.

**Comment by kevintelford** [comment]
Same as @Sevensidedmarble - for convenience we'd love to just run the console and track migrations from docker-compose, instead of introducing the Hasura CLI tool. The CLI tool itself has its uses obviously, but for some colleagues and responsibilities it'd be great to have one less boundary here.

Regardless, great work on this and thanks! Super useful service.

**Comment by Sevensidedmarble** [comment]
It looks like #3570 is pretty close, will that add the feature we want? And if so what's the time table on merging?

**Comment by shahidhk** [comment]
@Sevensidedmarble Yes, #3570 will address this issue. We're blocked on some console related changes for this to go in. Expecting to be in 1.3 or 1.4 release.

**Comment by vorpus** [comment]
> for convenience we'd love to just run the console and track migrations from docker-compose, instead of introducing the Hasura CLI tool. 

@kevintelford What is your current workaround for this? Make migrations on a local instance and sync with version control?

**Comment by plmercereau** [comment]
Hello, any idea on when I will be solved? Or a workaround while waiting?
Thanks

**Comment by peterldowns** [comment]
What's the status on this? We're running hasura inside kubernetes, and would like to avoid exposing our hasura endpoint publicly but still access the console. To access the console in such a situation, we would like to run `kubectl proxy` and then, with the graphql engine container running behind a service called `hasura`, run

```bash
hasura console \
  --endpoint 'http://localhost:8001/api/v1/namespaces/staging/services/hasura/proxy' \
  --admin-secret '<our-special-secret>'
```
Right now this works in that the cli program is able to auth correctly (`hasura metadata apply` works correctly), but the console is not actually accessible in the browser due to cross origin issues. It seems the that the browser console, running on domain `http://localhost:9695/`, attempts to make network requests directly to the endpoint on the host `http://localhost:8001`, which gets blocked with CORS issues:

![image](https://user-images.githubusercontent.com/824173/97199104-001eb100-1786-11eb-8be3-5fdcdfa2a3d3.png)

I see that from above the plan to proxy requests through the cli (#1440) was abandoned in favor of #3570, which also seems to have been abandoned as of July of this year. What do you recommend we do? I would really like to not publicly expose our hasura instance if at all possible (in our own webapp we proxy requests to it through our backend).

**Comment by kamac** [comment]
If you're running the console in a docker container, you can work around it by installing `socat` and running:

```sh
socat TCP-LISTEN:8080,fork TCP:host.docker.internal:8080 & hasura console --console-port 8081
```

Assuming you published port `8080` from `graphql-engine`, this will let the console communicate with the engine on `localhost:8080`. You'll be able to access the console from the browser on `http://localhost:8081`.

**Comment by elitan** [comment]
I'm also searching for a solution to this.

**Comment by bernatvadell** [comment]
> If you're running the console in a docker container, you can work around it by installing `socat` and running:
> 
> ```shell
> socat TCP-LISTEN:8080,fork TCP:host.docker.internal:8080 & hasura console --console-port 8081
> ```
> 
> Assuming you published port `8080` from `graphql-engine`, this will let the console communicate with the engine on `localhost:8080`. You'll be able to access the console from the browser on `http://localhost:8081`.
> 
> That's my temporary workaround.

Based on your example and with some modifications it has worked for me.
My directory structure for hasura console is as follows:

- metadata
- migrations
- config.yaml
- Dockerfile
 ```dockerfile
FROM ubuntu:20.04

RUN apt-get update -y && apt-get install -y curl socat

RUN curl -L https://github.com/hasura/graphql-engine/raw/stable/cli/get.sh | bash

WORKDIR /usr/src/hasura

COPY . .

RUN chmod +x ./start.sh

CMD "./start.sh"
 ```
- start.sh
```bash
#!/bin/bash

cd /usr/src/hasura || {
    echo "Hasura folder '/usr/src/hasura' not found"
    exit 1
}

# temporal fix to workaround: https://github.com/hasura/graphql-engine/issues/2824#issuecomment-801293056
socat TCP-LISTEN:8080,fork TCP:ext_graphql:8080 &
socat TCP-LISTEN:9695,fork,reuseaddr,bind=svc_hasura_console TCP:127.0.0.1:9695 &
socat TCP-LISTEN:9693,fork,reuseaddr,bind=svc_hasura_console TCP:127.0.0.1:9693 &
{
    # Apply migrations
    hasura migrate apply || exit 1

    # Apply metadata changes
    hasura metadata apply || exit 1

    # Run console if specified
    if [[ -v HASURA_RUN_CONSOLE ]]; then
        echo "Starting console..."
        hasura console --log-level DEBUG --address "127.0.0.1" --no-browser || exit 1
    else
        echo "Started without console"
        tail -f /dev/null
    fi
}
```
- docker-compose: 
```yml
version: "3.7"

volumes:
  database-data:

services:
  ext_db:
    image: postgres:12
    volumes:
      - database-data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 5s
      timeout: 5s
      retries: 5

  ext_graphql:
    image: hasura/graphql-engine:v1.3.3
    ports:
      - "8080:8080"
    depends_on:
      ext_db:
        condition: service_healthy
    environment:
      HASURA_GRAPHQL_DATABASE_URL: "postgres://postgres:${POSTGRES_PASSWORD}@ext_db:5432/postgres"
      HASURA_GRAPHQL_ENABLE_CONSOLE: "false"
      HASURA_GRAPHQL_DEV_MODE: "true"
      HASURA_GRAPHQL_ENABLED_LOG_TYPES: startup, http-log, webhook-log, websocket-log, query-log
      HASURA_GRAPHQL_ADMIN_SECRET: ${HASURA_GRAPHQL_ADMIN_SECRET}
    healthcheck:
      test: ["CMD-SHELL", "wget -qO- http://localhost:8080/healthz || exit 1"]
      interval: 5s
      timeout: 3s
      retries: 5

  svc_hasura_console:
    build:
      context: .
    ports:
      - "9693:9693"
      - "9695:9695"
    environment:
      HASURA_GRAPHQL_ADMIN_SECRET: ${HASURA_GRAPHQL_ADMIN_SECRET}
      HASURA_GRAPHQL_ENDPOINT: http://localhost:8080
      HASURA_RUN_CONSOLE: "true"
    volumes:
      - ./metadata:/usr/src/hasura/metadata
      - ./migrations:/usr/src/hasura/migrations
    depends_on:
      ext_graphql:
        condition: service_healthy
```

Thansk!

**Comment by cies** [comment]
I foud this issue because I have the same problem. I dont want to run ANYTHING without docker, just to ensure all within the team are running exactly the same versions/configs/etc (across different OSes).

I'm creating a workaround now (thanks for the `socat` tip), but the docs can be clearer on this matter, or just fix it.

**Comment by cies** [comment]
Ok, here's my workaround. In `docker-compose.yml` I have:

```  hasura:
    image: hasura/graphql-engine:v2.0.0-alpha.7.cli-migrations-v3
    ports:
      - "8080:8080"
      - "9695:9695"
      - "9693:9693"
    volumes:
      - ./hasura:/hasura
      - ./hasura/migrations:/hasura-migrations
      - ./hasura/metadata:/hasura-metadata
    depends_on:
      - postgres
    restart: always
    environment:
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:oijwqoiwjeqoijewq219827398321@postgres:5432/postgres
      # enable the console served by server
      HASURA_GRAPHQL_ENABLE_CONSOLE: "true" # set to "false" to disable console
      # enable debugging mode. It is recommended to disable this in production
      HASURA_GRAPHQL_DEV_MODE: "true"
      HASURA_GRAPHQL_ENABLED_LOG_TYPES: startup, http-log, webhook-log, websocket-log, query-log
      # uncomment next line to set an admin secret
      HASURA_GRAPHQL_ADMIN_SECRET: "wqoirewoi19287321987"
```

And I keep a `Makefile` with all commands I dont want to remember, which contains the following task:

```
.PHONY: hasura-console
hasura-console: ## 
        @docker-compose exec hasura bash -c "apt-get install -y socat; cd /hasura; \
                socat TCP-LISTEN:8080,fork TCP:hasura:8080 & \
                socat TCP-LISTEN:9695,fork,reuseaddr,bind=hasura TCP:127.0.0.1:9695 & \
                socat TCP-LISTEN:9693,fork,reuseaddr,bind=hasura TCP:127.0.0.1:9693 & \
                hasura-cli console --log-level DEBUG --address "127.0.0.1" --no-browser || exit 1 "

```

**Comment by cies** [comment]
Somethign else (I could make a new ticket for it), it seems Hasura has a naming issue. In the env vars I see `HASURA_GRAPHQL_ENABLE_CONSOLE` and this is not the same console as the one started with `hasura-cli console`. It is naming issues like this that make a product a lot harder to approach by newcomers.

I was confused too... It took me a while to know that `console` can mean different things in different contexts and that I needed the `console` to get migrations to work (and not the `console` (the other console), which does not save migrations).

**Comment by alecf** [comment]
I have pretty hacky but functional workaround, which is to edit your /etc/hosts to include all the various hosts you need. 

I have:
1) an ec2 linux development environment which runs docker i
  * the `app` container points to `graphql-engine` which is the hasura container, and references it in `docker-compose` as `http://graphql-engine:8080`
  * I then defined a separate `hasura-console` container with `docker-compose` that also talks to `http://graphql-engine:8080` but runs the console:
    ```yaml
    hasura-console:
      image: hasura/graphql-engine:v1.3.3.cli-migrations-v2
      # Note the entrypoint is listening on the "address" of hasura-console, which maps to the
      # IP address of *this* container!
      entrypoint: hasura-cli console --address hasura-console
      working_dir: /hasura
      env_file:
        - .env
      volumes:
        - ./hasura:/hasura
      environment:
        HASURA_GRAPHQL_ENDPOINT: http://graphql-engine:8080
      ports:
        - '9695:9695'
        - '9693:9693'
     ```
2) a mac laptop which runs chrome. On this machine I added a line to my /etc/hosts:
   > Replace 1.2.3.4 with your remote development environment IP address
   ```
   1.2.3.4 my-app graphql-engine hasura-console
   ```

Then from my laptop I can access `http://hasura-console:9695` - and the web app itself will attempt to connect to `http://hasura-console:9693` as well as `http://graphql-engine:8080` - but they all ultimately point to the same machine on different ports! This lets me do schema migrations in the hasura console, recording the migrations to a locally mounted filesystem on the development box

**Comment by Californian** [comment]
Here's what we use in case it's helpful. I didn't include the traefik or dnsmasq config because the latter only works on linux (mac dns is a bit of a nightmare) so just ignore the labels.

It's organized as follows:

```sh
├── development-stack
│   ├── docker-compose.yml
│   ├── .env
│   ├── README.md
│   └── scripts
└── hasura-console
    ├── app
    │   ├── config.yaml
    │   ├── metadata
    │   ├── migrations
    │   └── seeds
    ├── Dockerfile
    ├── README.md
    ├── scripts -> ../development-stack/scripts
    └── start.sh
```

`Dockerfile`:
```Dockerfile
FROM debian:buster-slim

RUN apt-get update -y && apt-get install -y curl socat
RUN curl -L -o /usr/local/bin/hasura https://github.com/hasura/graphql-engine/releases/download/v2.0.0-beta.2/cli-hasura-linux-amd64
RUN chmod +x /usr/local/bin/hasura

WORKDIR /usr/src/hasura

COPY . .

RUN chmod +x ./start.sh

CMD "./start.sh"
```

`start.sh`:
```sh
#!/bin/bash

HASURA_FOLDER=/usr/src/hasura/app
cd $HASURA_FOLDER || {
    echo "Hasura folder '$HASURA_FOLDER' not found"
    exit 1
}

# Workaround for https://github.com/hasura/graphql-engine/issues/2824#issuecomment-801293056
socat TCP-LISTEN:8080,fork TCP:hasura-engine:8080 &
socat TCP-LISTEN:9695,fork,reuseaddr,bind=hasura-console TCP:127.0.0.1:9695 &
socat TCP-LISTEN:9693,fork,reuseaddr,bind=hasura-console TCP:127.0.0.1:9693 &
{
    # Apply migrations
    hasura migrate apply --database-name=default || exit 1

    # Apply metadata changes
    hasura metadata apply || exit 1

    # Run console if specified
    if [[ -v HASURA_RUN_CONSOLE ]]; then
        echo "Starting console..."
        hasura console --log-level DEBUG --address "127.0.0.1" --no-browser || exit 1
    else
        echo "Started without console"
        tail -f /dev/null
    fi
}
```

`docker-compose.yml` (`traefik` and `dnsmasq` services omitted because they aren't relevant):
```
version: '3.8'
services:
  db:
    image: postgres:13.3-alpine
    ports:
      - "5432:5432"
    volumes:
      - db_data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_DB: app
    networks:
      - db
    restart: unless-stopped
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 2s
      timeout: 2s
      retries: 10

  hasura-engine:
    image: hasura/graphql-engine:v2.0.0-beta.2
    ports:
      - "8080:8080"
    depends_on:
      db:
        condition: service_healthy
    environment:
      HASURA_GRAPHQL_DATABASE_URL: postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:5432/app
      HASURA_GRAPHQL_ENABLE_CONSOLE: "false"
      HASURA_GRAPHQL_DEV_MODE: "true"
      HASURA_GRAPHQL_ENABLED_LOG_TYPES: startup, http-log, webhook-log, websocket-log, query-log
      HASURA_GRAPHQL_ADMIN_SECRET: ${HASURA_GRAPHQL_ADMIN_SECRET}
      HASURA_GRAPHQL_ENABLE_TELEMETRY: "false"
      HASURA_GRAPHQL_JWT_SECRET: ${HASURA_GRAPHQL_JWT_SECRET}
    networks:
      - web
      - db
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=web"
      - "traefik.http.routers.hasura-engine.rule=Host(`hasura-engine.app.test`)"
    tty: true
    restart: unless-stopped
    healthcheck:
      test: ["CMD-SHELL", "apt-get update -y && apt-get install -y curl && curl --fail http://localhost:8080/healthz || exit 1"]
      interval: 5s
      timeout: 3s
      retries: 5

  hasura-console:
    build:
      context: ../../hasura-console
    ports:
      - "9693:9693"
      - "9695:9695"
    environment:
      HASURA_GRAPHQL_DATABASE_URL: postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:5432/app
      HASURA_GRAPHQL_ADMIN_SECRET: ${HASURA_GRAPHQL_ADMIN_SECRET}
      HASURA_GRAPHQL_ENDPOINT: http://127.0.0.1:8080
      HASURA_RUN_CONSOLE: "true"
    volumes:
      - ../../hasura-console:/usr/src/hasura
    networks:
      - web
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=web"
      - "traefik.http.services.hasura-console.loadbalancer.server.port=9695"
      - "traefik.http.routers.hasura-console.rule=Host(`hasura-console.app.test`)"
    depends_on:
      hasura-engine:
        condition: service_healthy
```

**Comment by radicand** [comment]
For those that are still following this, I use VSCode's devcontainers feature, and have the following scripts/features to enable it to work as you'd expect:

In the Dockerfile for the dev container, I have the following:

```Dockerfile
FROM hasura/graphql-engine:v2.3.1 AS consoleassets
... your Dockerfile, including FROM line ...
COPY --from=consoleassets /srv/console-assets/ /srv/console-assets/
RUN curl -L https://github.com/hasura/graphql-engine/raw/stable/cli/get.sh | bash
... other things you need ...
```

Then I have a node script (my base image includes node, install if yours doesn't):

```js
const express = require('express');
const morgan = require('morgan');
const cors = require('cors');
const { createProxyMiddleware } = require('http-proxy-middleware');

// Configuration
const PORT = 8080;
const HOST = 'localhost';
const API_SERVICE_URL = 'http://graphql-engine:8080';

// Create Express Server
const app = express();
app.use(
  cors({
    origin: [
      'http://localhost:8080',
      'http://localhost:9695',
      'http://localhost:9696',
    ],
  })
);

// Logging
app.use(morgan('dev'));

app.use(
  '/console/assets/',
  express.static('/srv/console-assets/', {
    setHeaders: function (res, path, stat) {
      if (path.endsWith('.gz')) {
        res.set('Content-Encoding', 'gzip');
        res.set(
          'Content-Type',
          path.endsWith('.js.gz') ? 'application/javascript' : 'text/css'
        );
      }
    },
  })
);

// Proxy endpoints
app.use(
  '/v*/',
  createProxyMiddleware({
    target: API_SERVICE_URL,
    changeOrigin: true,
  })
);
app.use(
  '/',
  createProxyMiddleware({
    target: `http://localhost:9695`,
    changeOrigin: true,
  })
);

// Start the Proxy
app.listen(PORT, HOST, () => {
  console.log(`Starting Proxy at ${HOST}:${PORT}`);
});
```

Choose how you want to run that script, I copy it into the image and execute a `npm i` to install the dependencies it needs, and then start it on container startup via `devcontainer.json` param `"postStartCommand": "node /home/vscode/.local/nodeproxy.js",` (assuming this is where you saved it and installed the npm dependencies).

Now, when you start your devcontainer environment, all you need to do is navigate to the workspace folder that has your metadata and migrations, and execute `hasura console --endpoint http://localhost:8080 --use-server-assets --no-browser --skip-update-check` - this will work offline as well (you know, when you want to work on a airplane for instance) because you copied in the console assets in the Dockerfile and the proxy script maps it.

Hope this helps someone else!

**Comment by corysimmons** [comment]
Does anyone have an updated version of this that works? This is super frustrating...

**Comment by jorroll** [comment]
For folks who are finding their way to this issue out of a desire to integrate Hasura in a [devcontainer](https://code.visualstudio.com/docs/devcontainers/containers), here's a simple solution involving a minor compromise. 

The magic bit that makes it work is `network_mode: host` for the devcontainer (i.e. `workspace`) service. By setting the devcontainer service to `network_mode: host`, it will share the same network as the host computer meaning that `localhost` in the devcontainer is the same as `localhost` on the host computer. In my case, this was an acceptable (no brainer) compromise since I was already exposing all of my devcontainer's ports on the host computer. 

It is expected that the hasura-cli will be installed in the devcontainer (i.e. `workspace`) service.

### Docker compose file

```yaml
# .devcontainer/docker-compose.yaml

# README for the Hasura portion of this file
# https://github.com/hasura/graphql-engine/tree/stable/install-manifests/docker-compose-pgadmin

# Explanation of networking in Docker Compose
# https://docs.docker.com/compose/networking/
# and also https://hasura.io/docs/latest/deployment/deployment-guides/docker/
# - Different services can access each other via their name.
#   E.g. `curl http://pgadmin:80` from vscode's terminal

version: "3.8"
services:
  postgres:
    image: postgres:14
    volumes:
      - db_storage:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: postgrespassword
  pgadmin:
    image: dpage/pgadmin4
    depends_on:
      - postgres
    ports:
      - 5050:80
    ## you can change pgAdmin default username/password with below environment variables
    environment:
      PGADMIN_DEFAULT_EMAIL: pgadmin@example.com
      PGADMIN_DEFAULT_PASSWORD: admin
  graphql-engine:
    # Update this in tandom with the `hasura-cli` NPM package in the
    # repo root package.json
    image: hasura/graphql-engine:v2.15.1
    ports:
      - 8080:8080
    depends_on:
      - "postgres"
    restart: always
    environment:
      ## postgres database to store Hasura metadata
      HASURA_GRAPHQL_METADATA_DATABASE_URL: postgres://postgres:postgrespassword@postgres:5432/postgres
      ## this env var can be used to add the above postgres database to Hasura as a data source. this can be removed/updated based on your needs
      PG_DATABASE_URL: postgres://postgres:postgrespassword@postgres:5432/postgres
      ## enable the console served by server
      HASURA_GRAPHQL_ENABLE_CONSOLE: "false" # set to "false" to disable console
      ## enable debugging mode. It is recommended to disable this in production
      HASURA_GRAPHQL_DEV_MODE: "true"
      HASURA_GRAPHQL_ENABLED_LOG_TYPES: startup, http-log, webhook-log, websocket-log, query-log
      HASURA_GRAPHQL_EXPERIMENTAL_FEATURES: naming_convention
      HASURA_GRAPHQL_DEFAULT_NAMING_CONVENTION: graphql-default
      ## uncomment next line to run console offline (i.e load console assets from server instead of CDN)
      HASURA_GRAPHQL_CONSOLE_ASSETS_DIR: /srv/console-assets
      ## uncomment next line to set an admin secret
      HASURA_GRAPHQL_ADMIN_SECRET: myadminsecretkey
      ## uncomment next line to set an JWT secret
      # HASURA_GRAPHQL_JWT_SECRET: yoursecret
      # HASURA_GRAPHQL_UNAUTHORIZED_ROLE: anonymous
  workspace:
    build:
      context: ./
      dockerfile: Dockerfile
      args:
        # Update 'VARIANT' to pick a Node version: 16, 14, 12.
        # Append -bullseye or -buster to pin to an OS version.
        # Use -bullseye variants on local on arm64/Apple Silicon.
        VARIANT: 16-bullseye
    volumes:
      - ..:/workspace:cached
    depends_on:
      - postgres
      - graphql-engine
    # We need to use "host" network mode in order to use Hasura CLI within
    # the devcontainer due to this issue:
    # https://github.com/hasura/graphql-engine/issues/2824
    network_mode: host
    # Overrides default command so things don't shut down after the process ends.
    command: /bin/sh -c "while sleep 1000; do :; done"
volumes:
  db_storage:
```

### Devcontainer.json

```json
// .devcontainer/devcontainer.json

// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
// https://github.com/microsoft/vscode-dev-containers/tree/v0.224.2/containers/typescript-node
{
  "name": "my container",
  // The 'service' property is the name of the service for the container that VS Code should
  // use. Update this value and .devcontainer/docker-compose.yml to the real service name.
  "service": "workspace",

  // The optional 'workspaceFolder' property is the path VS Code should open by default when
  // connected. This is typically a file mount in .devcontainer/docker-compose.yml
  "workspaceFolder": "/workspace",

  // Update the 'dockerComposeFile' list if you have more compose files or use different names.
  // The .devcontainer/docker-compose.yml file contains any overrides you need/want to make.
  "dockerComposeFile": ["docker-compose.yaml"],

  // Set *default* container specific settings.json values on container create.
  "settings": {
    "terminal.integrated.shell.linux": "/bin/zsh"
  },

  // Add the IDs of extensions you want installed when the container is created.
  // "extensions": [],

  // Use 'postCreateCommand' to run commands after the container is created.
  // "postCreateCommand": "",

  // Comment out to connect as root instead. More info: https://aka.ms/vscode-remote/containers/non-root.
  "remoteUser": "node",

  // "features": {
  //   "git": "latest",
  //   "java": "lts"
  // }
}
```

### DockerFile

```dockerfile
# .devcontainer/DockerFile

# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.224.2/containers/typescript-node/.devcontainer/base.Dockerfile

# [Choice] Node.js version (use -bullseye variants on local arm64/Apple Silicon): 16, 14, 12, 16-bullseye, 14-bullseye, 12-bullseye, 16-buster, 14-buster, 12-buster
ARG VARIANT="16-bullseye"
FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-${VARIANT}

# [Optional] Uncomment this section to install additional OS packages.
# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
#    && apt-get -y install --no-install-recommends zsh-antigen

# [Optional] Uncomment if you want to install an additional version of node using nvm
# ARG EXTRA_NODE_VERSION=10
# RUN su node -c "source /usr/local/share/nvm/nvm.sh && nvm install ${EXTRA_NODE_VERSION}"

# [Optional] Uncomment if you want to install more global node packages
# RUN su node -c "npm install -g gulp-cli"
```

**Comment by manasag** [comment]
CLI `hasura console` command supports `--api-host` argument which can be used to set a non localhost address of the hasura server (see this [documentation](https://github.com/hasura/graphql-engine/pull/7519)). We think this should support the docker container usecase. We are going to test this internally and provide an update in some time.

**Comment by manasag** [comment]
Hi all, we have recently merged change which will allow configuring hasura endpoint separately in cli, thus enabling CLI Console in a container. More details to be shared by the next release 2.18

**Comment by m-Bilal** [comment]
Hi all, we have added a new flag `--console-hge-endpoint` to the `hasura console` command, which will allow you to set the HGE (Hasura GraphQL Engine) endpoint for the console. This variable not being set correctly is what was causing the console to not be able to access the HGE APIs when the CLI was inside of a Docker container. [Docs](https://hasura.io/docs/latest/hasura-cli/commands/hasura_console/) have more info, but the gist is: this flag will have the same address as the `--address` flag. `--address` specifies where the CLI should reach the HGE, `--console-hge-endpoint` will specify where the console should reach the HGE. Example (assuming HGE Docker container exposes port 8080):
```
hasura console --endpoint http://host.docker.internal:8080/ --no-browser --address 0.0.0.0 --console-hge-endpoint http://0.0.0.0:8080/
```

**Comment by 27medkamal** [comment]
I am happy to confirm that this works with the latest version of hasura. Here's a working example of a `docker-compose.yml` file which should work with a standard hasura project created using `hasura init`

```
version: '3.9'
services:
  db:
    image: postgres:14.2
    environment:
      POSTGRES_INITDB_ARGS: --data-checksums
      POSTGRES_DB: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_USER: postgres

  hasura:
    image: hasura/graphql-engine:v2.19.0.cli-migrations-v3
    volumes:
      - ./migrations:/hasura-migrations
      - ./metadata:/hasura-metadata
    depends_on:
      - db
    environment:
      HASURA_GRAPHQL_ENABLE_CONSOLE: 'false'
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:postgres@db:5432/postgres
      HASURA_GRAPHQL_ADMIN_SECRET: secret
    ports:
      - '8080:8080'

  hasura_console:
    image: hasura/graphql-engine:v2.19.0.cli-migrations-v3
    volumes:
      - .:/app
    working_dir: /app
    command: hasura-cli console
      --no-browser
      --endpoint http://hasura:8080
      --admin-secret secret
      --address 0.0.0.0
      --console-port 9695
      --api-port 9693
      --console-hge-endpoint http://localhost:8080
    ports:
      - '9695:9695'
      - '9693:9693'
    depends_on:
      - hasura
    environment:
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:postgres@db:5432/postgres

```

**Comment by samfweb** [comment]
Thanks @27medkamal, that was a huge help.

The console kept starting before hasura was ready.

Adding a healthcheck worked for me:
```yaml
version: "3.9"
services:
  db:
    image: postgres:14.2
    environment:
      POSTGRES_INITDB_ARGS: --data-checksums
      POSTGRES_DB: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_USER: postgres

  hasura:
    image: hasura/graphql-engine:v2.19.0.cli-migrations-v3
    volumes:
      - ./server/hasura/migrations:/hasura-migrations
      - ./server/hasura/metadata:/hasura-metadata
    depends_on:
      - db
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8080/v1/version"]
      interval: 10s
      timeout: 5s
      retries: 5
    environment:
      HASURA_GRAPHQL_ENABLE_CONSOLE: "false"
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:postgres@db:5432/postgres
      HASURA_GRAPHQL_ADMIN_SECRET: secret
    ports:
      - "8080:8080"

  hasura_console:
    image: hasura/graphql-engine:v2.19.0.cli-migrations-v3
    volumes:
      - ./server/hasura:/app
    working_dir: /app
    command: hasura-cli console
      --no-browser
      --endpoint http://hasura:8080
      --admin-secret secret
      --address 0.0.0.0
      --console-port 9695
      --api-port 9693
      --console-hge-endpoint http://localhost:8080
    ports:
      - "9695:9695"
      - "9693:9693"
    depends_on:
      hasura:
        condition: service_healthy
    environment:
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:postgres@db:5432/postgres
```

**Comment by samfweb** [comment]
Thanks @27medkamal for this. 

I can't seem to get the above working. The console will launch but it can't find the server. The following error is thrown inside the console. 

```
Schema introspection query failed
x-hasura-admin-secret/x-hasura-access-key required, but not found
```

Any idea why this may be?

I also had to add a health check to prevent the console from starting too early:
```yaml
version: "3.9"
services:
  db:
    image: postgres:14.2
    environment:
      POSTGRES_INITDB_ARGS: --data-checksums
      POSTGRES_DB: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_USER: postgres

  hasura:
    image: hasura/graphql-engine:v2.19.0.cli-migrations-v3
    volumes:
      - ./server/hasura/migrations:/hasura-migrations
      - ./server/hasura/metadata:/hasura-metadata
    depends_on:
      - db
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8080/v1/version"]
      interval: 10s
      timeout: 5s
      retries: 5
    environment:
      HASURA_GRAPHQL_ENABLE_CONSOLE: "false"
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:postgres@db:5432/postgres
      HASURA_GRAPHQL_ADMIN_SECRET: secret
    ports:
      - "8080:8080"

  hasura_console:
    image: hasura/graphql-engine:v2.19.0.cli-migrations-v3
    volumes:
      - ./server/hasura:/app
    working_dir: /app
    command: hasura-cli console
      --no-browser
      --endpoint http://hasura:8080
      --admin-secret secret
      --address 0.0.0.0
      --console-port 9695
      --api-port 9693
      --console-hge-endpoint http://localhost:8080
    ports:
      - "9695:9695"
      - "9693:9693"
    depends_on:
      hasura:
        condition: service_healthy
    environment:
      HASURA_GRAPHQL_DATABASE_URL: postgres://postgres:postgres@db:5432/postgres

```

**Comment by 27medkamal** [comment]
@samfweb You're on the right track. I've added the healthchecks to my [comment above](https://github.com/hasura/graphql-engine/issues/2824#issuecomment-1436457924) which should now work. I initially tested it with the healthchecks but thought I'd remove them in order not to bloat the snippet, but I've added them and it should now work.

**Comment by 27medkamal** [comment]
@samfweb Also I'd suggest running this with a blank hasura project created using `hasura init` in case your current project setup is affecting this docker-compose setup

**Comment by m-Bilal** [comment]
@samfweb could you please check if the admin secret is correct? The error seems to be hinting towards it: `x-hasura-admin-secret/x-hasura-access-key required, but not found`

**Comment by samfweb** [comment]
@27medkamal I've created a completely blank project and I'm still running into a heap of errors. It seems like some of the flags aren't making their way into the console properly.

In the screenshot below, you can see it's trying to reach `http://undefined/v1/graphql` instead of `http://hasura/v1/graphql`

<details><summary>Error snapshot</summary>

![image](https://user-images.githubusercontent.com/44712417/220799099-9dd23ef6-7212-4758-92e6-89d7034ce364.png)

</details>

Thanks for the tip @m-Bilal, somehow my local storage was overwriting the request headers. Clearing my local storage helped fix that error.

**Comment by samfweb** [comment]
@m-Bilal It looks like `--console-hge-endpoint` isn't working correctly

**Comment by 27medkamal** [comment]
@samfweb You are right. It seems like it's working as you're browsing but when you inspect the console and the network tab, there are some issues with a few endpoint calls. @m-Bilal thoughts?

**Comment by 27medkamal** [comment]
@samfweb @m-Bilal After a bit of testing, it seems that those issues happen even if the conosle is started from outside docker. It's an issue nonetheless, but it doesn't seem specific to running the hasura console within docker

**Comment by m-Bilal** [comment]
@27medkamal You're right, the errors are unrelated to running the console from inside a container.

@samfweb, could you please try using the console from another browser and checking if it works fine?

**Comment by semiautomatix** [comment]
@m-Bilal thanks for the update on this.

I have a use case where we're using Gitpod. As such the publically accessible addresses are all port 443, e.g. `https://8080-saretail-retmembersv2-abc123.ws-us90.gitpod.io/` or `https://9693-saretail-retmembersv2-abc123.ws-us90.gitpod.io/`. We would specifically need to specify the a different address for the api endpoints.

**Comment by samfweb** [comment]
Apologies for the delayed response @m-Bilal 

I tried it on both Chrome and Brave with the same result.

**Comment by andoks** [comment]
Running hasura console in a docker-container on windows (using docker-desktop in WSL2 configuration) has multiple issues for us. Passing `--console-hge-endpoint=http://0.0.0.0:8080` causes the console to not work at all (showing a loading screen that never goes away). By setting `--console-hge-endpoint=http://localhost:8080` we got the "API" tab to work somewhat, to the point where we can at least create and execute queries in the GraphiQL window, but trying to open any of the other tabs ("DATA", "ACTIONS", "REMOTE SCHEMAS" & "EVENTS"). 

It seems that maybe the reason for why these tabs does not work is that the console tries to fetch data from http://0.0.0.0:9693/apis/migrate/settings, but this yields an `net::ERR_ADDRESS_INVALID` according to Google Chromes developer console.

**Comment by badis** [comment]
@andoks  I am facing exactly the same issue, did you figure out how to fix it ? 

![issue](https://user-images.githubusercontent.com/3008693/230684541-56b7a982-3ac4-4b04-be6e-133d873a0e1b.PNG)

**Comment by andoks** [comment]
@badis No, the GraphiQL tab worked on Windows for us, while the rest of the tabs also works on Linux, so we are just making do with that for now. I hope it eventually gets fixed though, or else we might have to consider a different approach than putting the hasura cli in a docker image.

**Comment by dgbau** [comment]
I might have a solution! I have been running into this same/similar issue, trying to launch the console from a VS Code SSH session. I found that VSC was porting 9695 automatically when the command was run, but was not porting 9693. I manually added that port to the forwarding, and now the call to `http://localhost:9693/apis/migrate/settings` succeeds. I am not using any flags in the console command besides `--project`, `--secret` and `--endpoint`. 

It seems like this is new, either 9693 was getting automatically added to forwarded ports before, or the calls to the above URL were on 9695.

**Comment by m-Bilal** [comment]
Sorry folks, I lost track of this. I'll be looking into these and getting back to you in the next couple weeks.

**Comment by GiamBoscaro** [comment]
> Running hasura console in a docker-container on windows (using docker-desktop in WSL2 configuration) has multiple issues for us. Passing `--console-hge-endpoint=http://0.0.0.0:8080` causes the console to not work at all (showing a loading screen that never goes away). By setting `--console-hge-endpoint=http://localhost:8080` we got the "API" tab to work somewhat, to the point where we can at least create and execute queries in the GraphiQL window, but trying to open any of the other tabs ("DATA", "ACTIONS", "REMOTE SCHEMAS" & "EVENTS").
> 
> It seems that maybe the reason for why these tabs does not work is that the console tries to fetch data from http://0.0.0.0:9693/apis/migrate/settings, but this yields an `net::ERR_ADDRESS_INVALID` according to Google Chromes developer console.

I am facing the exact same problem, but in Linux. Using Ubuntu 18:04.

Hasura Console: 2.28.0
GraphQL Engine: 2.22.0

```shell
# entrypoint.sh
hasura console \
    --endpoint $HASURA_GRAPHQL_ENDPOINT \ # http://hasura:8080    graphq engine container
    --admin-secret $HASURA_GRAPHQL_ADMIN_SECRET \
    --address 0.0.0.0 \
    --console-hge-endpoint http://localhost:8080 \ # not sure localhost is right here but it is the only way to run the console, otherwise the container crashes
    --log-level $HASURA_CLI_LOG_LEVEL \
    --no-browser || exit 1
```

**Comment by ranjan-purbey** [comment]
You can use `network_mode: host` if running on linux hosts like so:
```yaml
version: "3.8"
services:
  graphql-engine:
    image: hasura/graphql-engine:v2.30.1-ce.cli-migrations-v3
    ports:
      - 8080:8080
    restart: no
    environment:
      ## postgres database to store Hasura metadata
      HASURA_GRAPHQL_METADATA_DATABASE_URL: <POSTGRES_CONNECTION_STRING>
      ## this env var can be used to add the above postgres database to Hasura as a data source. this can be removed/updated based on your needs
      PG_DATABASE_URL: <POSTGRES_CONNECTION_STRING>
      ## enable the console served by server
      HASURA_GRAPHQL_ENABLE_CONSOLE: "false" # set to "false" to disable console
      ## enable debugging mode. It is recommended to disable this in production
      HASURA_GRAPHQL_DEV_MODE: "true"
      HASURA_GRAPHQL_ENABLED_LOG_TYPES: startup, http-log, webhook-log, websocket-log, query-log
      ## uncomment next line to run console offline (i.e load console assets from server instead of CDN)
      HASURA_GRAPHQL_CONSOLE_ASSETS_DIR: /srv/console-assets
      ## uncomment next line to set an admin secret
      HASURA_GRAPHQL_ADMIN_SECRET: myadminsecretkey
    healthcheck:
      test: nc -z localhost 8080
      interval: 5s
      timeout: 5s
      start_period: 10s
  console:
    depends_on:
      graphql-engine:
        condition: service_healthy
    image: hasura/graphql-engine:v2.30.1-ce.cli-migrations-v3
    network_mode: host
    restart: no
    volumes:
      - ./hasura:/hasura
    working_dir: /hasura
    entrypoint: ["sh", "-c", "hasura-cli console --no-browser"]
    environment:
      HASURA_GRAPHQL_ADMIN_SECRET: myadminsecretkey
```
---
I think the problem can be solved by updating the following: https://github.com/hasura/graphql-engine/blob/4d86cd3a62e693b4257e2812ce71ae02ce226e19/cli/commands/console.go#L139
```diff
- apiServer, err := console.NewAPIServer(o.APIHost.Host, o.APIPort, o.EC)
+ apiServer, err := console.NewAPIServer(o.Address, o.APIPort, o.EC)
```
I was unable to build the repo locally, so cannot verify the fix. Can someone try it?

**Comment by andoks** [comment]
@ranjan-purbey: Do you know if this will work when running on docker-desktop for windows, which runs the docker containers on what is essentially a linux VM?

**Comment by ranjan-purbey** [comment]
@andoks I have updated my comment to work without having to use the 'host' network driver. Now it should work on all docker distributions including windows

**Comment by ranjan-purbey** [comment]
@andoks can you please share the compose file you used for this?

**Comment by andoks** [comment]
@ranjan-purbey do you mean the full setup? Unfortunately I cannot share that, and will have to make minimal reproduction at some point if you want me to reproduce it. This is the service file with some redactions that I used for the console testing it with my system.

But is it expected of the hasura console feature that the --endpoint and --console-hge-endpoint changes where the console attempts to retrieve the static assets from? I would assume you could reproduce that just by changing the value you pass to --console-hge-endpoint  in your own example, and if when you open the console in your browser and open the developer console, the console does not attempt to fetch resources from the address you passed to --console-hge-endpoint, then something is wrong with my setup, but if it does so on your setup too, then it might be a bug unless it is intended to to this (in which case I have misunderstood what the --console-hge-endpoint flag is used for entirely)

```yaml
version: "3.6"

# context and file paths are specified starting at PWD, and PWD is required to
# be repository root

services:
  console:
    depends_on:
      project-hasura:
        condition: service_healthy
    image: hasura/graphql-engine:v2.30.1-ce.cli-migrations-v3
    ports:
      - 9695:9695
      - 9693:9693
    volumes:
      - ${PWD}/hasura-server:/hasura
    working_dir: /hasura
    entrypoint:
      [
        'sh',
        '-c',
        'hasura-cli console --no-browser --address `hostname -i` --endpoint http://hasura:8080 --console-hge-endpoint http://192.168.202.102:9695',
      ]
    environment:
      HASURA_GRAPHQL_ADMIN_SECRET: "<some secret>"
```

**Comment by LeulAria** [comment]
what's the update here ?
